### DISCUSSION TITLE
[Answered] Proposal: Logger library

### DISCUSSION THREAD

**Discussion by zjeffer** [nuttyartist/notes#486] [Ideas]
Currently, we can only log (to console or file) by including `#iostream` and using std::cout. I propose we use a logger so we can easily log to a file or to stdout. 

I'm a big fan of g3log: https://github.com/KjellKod/g3log. It allows you to define your own sinks (file, stdout, etc) and your own syntax & colors for every log level (INFO, DEBUG, ERROR, WARN, TRACE, etc...), including the file and linenumber where the log is called.

Here's an example in my connect4 deep RL engine project: https://github.com/zjeffer/connect4-deep-rl/tree/main/src/logger

This can be used like so:

```cpp
LINFO << "This is an INFO level log"; // no need to do std::endl, this is added in the sink.
LDEBUG << "DEBUG statement";
```

This would be incredibly useful for development & debugging purposes, and people could maybe pass a --debug argument to enable the DEBUG/TRACE level when reporting bugs/crashes?

**Comment by bjorn** [comment]
Since we're using Qt, why would we not just use Qt's logging features? It can be made to log to a file through [`qtInstallMessageHandler`](https://doc.qt.io/qt-5/qtglobal.html#qInstallMessageHandler), formatting and additional information can be customized using [`qSetMessagePattern` ](https://doc.qt.io/qt-5/qtglobal.html#qSetMessagePattern), it also supports info/debug/error/etc. levels and [custom  categories](https://doc.qt.io/qt-5/qloggingcategory.html).

We can already use it in similar fashion:
```c++
qInfo() << "This is an INFO level log"; // no need to do std::endl
qDebug() << "DEBUG statement";
```

**Reply by zjeffer** [reply]
Ah, I didn't know Qt had logging features. We can use that instead. Thanks.
